components/router/router.js

```diff
@@ -4,59 +4,46 @@
  *--------------------------------------------------------------------------------------------*/
 
 import components from 'coherent-gameface-components';
+import { HashHistory, BrowserHistory } from './history';
+import { Route } from './route';
 
 const WILDCARD = '**';
 const HOME = '/';
 
 class Router {
+    /**
+     * @typedef {Record<string, string|Router>} Routes
+     */
+
+    /**
+     * @param {Routes} routes - The routes map
+     * @param {HashHistory|BrowserHistory} history - The router history
+     * @param {any} onBeforeNavigation - Method that can be used for code execution before the next history navigation
+     */
     constructor(routes, history, onBeforeNavigation) {
         this.routes = routes;
 
         if (history) {
             this.history = history;
             if (onBeforeNavigation) this.history.onBeforeNavigation = onBeforeNavigation;
 
-            window.addEventListener('popstate', (event) => this.onPopState(event));
-
-            this.history.listen((current) => {
-                this.navigateTo(current);
-            });
+            this.history.addPopStateListener();
+            this.history.onHistoryChange = () => this.navigateTo(window.history.state.current);
         }
     }
 
     /**
-     * Called onpopstate event.
-     * Does nothing if there's no history or no history state.
-     * @param {Event} event - the current event object.
+     * Will clear the router data and remove the history listeners
      */
-    onPopState(event) {
-        if (!window.history || !window.history.state) return;
-
-        // onpopstate is triggered when back, go or forward is called and the event object
-        // does not hold information about which one of these methods triggered the event
-        // so we keep track of the currentRouteId, we set it ot each pushState and we compare the
-        // values on popstate to see if the user's going back or forward
-        const navDirection = this.getNavigationDirection(this.history.currentRouteId, event.state.id);
-        this.history.currentRouteId += navDirection;
-
-        if (!this.history.onBeforeNavigation) return this.navigateTo(window.history.state.current);
+    clear() {
+        if (this.history) {
+            this.history.removePopStateListener();
+            this.history.onBeforeNavigation = null;
+            this.history.onHistoryChange = null;
+            this.history = null;
```

Comment:
Should we also nullify the window.history?

Reply:
why we should do that? window.history is the global history object and it is read-only.